proposed/2022/NuGetExeWarnWhenSDKProjectsGetPacked.md

```diff
@@ -0,0 +1,99 @@
+```
+
+The command packages the single assembly, but does not generate the correct dependencies or any other metadata in the same way that the pack SDK would.
+
+Scenario 2: Multiple target framework SDK based projects, with .NET SDK installed on the machine.
+
+```console
+Attempting to build package from 'SDK.csproj'.
+MSBuild auto-detection: using msbuild version '17.4.0.37601' from 'C:\Program Files\Microsoft Visual Studio\2022\Preview\MSBuild\Current\Bin\amd64'.
+Error NU5012: Unable to find 'bin\Debug\SDK\bin\Debug\'. Make sure the project has been built.
+```
+
+NuGet.exe does not understand the concept of multitargetting and just merges the output paths.
+
+Scenario 3: SDK-based project, without the .NET SDK installed
+
+```console
+The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Code\Temp\SDK\SDK.csproj
+```
+
+The parsing fails early.
+
+### Proposal
+
+When someone attempts to run pack on an SDK-based project, NuGet will fail with the following error:
+
+```console
+Error NU5049: The `pack` command for SDK-style projects is not supported, use `dotnet pack` to pack this project instead. You may set the ENABLE_LEGACY_NUGETEXE_CSPROJ_PACK environment variable to revert to the previous packing behavior.
+```
+
+This *breaking change* will be announced via blog well before the NuGet.exe carrying the change ships.
+
+### Technical explanation
+
+As described above, whenever there's no SDK installed, there's simply no accurate way for NuGet to detect that a project is SDK-based.
+We can do some XML parsing, but it is likely to be error prone. `NuGet.exe pack` is not the only command that will fail, so the customer will likely figure out that they need to install the .NET SDK. After that the same scenarios with the installed .NET SDK will follow.
+
+To determine whether a project is `.NET SDK` based, we will evaluate the project (part of the normal pack operation), and check for the `UsingMicrosoftNETSDK` property being set. If the property is set, then the project is presumed to be .NET SDK based and will be blocked from packing.
+
+## Drawbacks
+
+- The proposal here would be `breaking` for customers, but it's more than likely that the scenario for which they used nuget.exe pack on an SDK based project never worked correctly.
+
+## Rationale and alternatives
+
+- Do nothing. This would mean just acknowledging that SDK-based projects support in NuGet.exe would not be added and not do anything to help migrate users to `dotnet.exe pack` and `msbuild /t:pack`.
+- Add full pack SDK support into NuGet.exe. While not impossible, this creates maintainability considerations. dotnet.exe has a more involved support for the SDK-based PackageReference projects, and as such is considered to be the `right tool for the job`.
+- Call `msbuild /t:pack` from NuGet.exe. This has the same considerations as the previos approach that `dotnet.exe` should be recommended as the right tool for the job.

```

Comment:
Why this won't work? The `dotnet pack` also calls `dotnet msbuild -t:Pack` although it's Core MSBuild. There shouldn't be any differences between Core/Full MSBuild. They are just targeting different runtime, that's all.

_There are proposals to MSBuild to break away from compat. But it's unlikely to happen anytime soon. May be after .NET Framework 4.9 or .NET 9_ 😇

Reply:
We thought quietly switching out the implementation doesn't really inform users that `dotnet` is what they should use with SDK based projects.

It's a 2 fold effort given that `dotnet` has extra commands for SDK projects that are only available there.

Reply:
True but if we want to inform users, we could say...

```
> nuget pack ...
Using NuGet Pack targets (msbuild -t:Pack)
```

```
> nuget pack ...
Using NuGet Pack CLI
```

We're only going to map available commands so as to not break existing CI/CD pipelines/workflows. With the above message people can decide when they want to move out of `nuget pack`. Besides you can call this out in the blog post. Sure, it might not have the same effect of seeing on the CLI on every invoke. Still a solid way of doing things properly.

Reply:
> True but if we want to inform users, we could say...

I don't think that'd be good enough given what we're trying to communicate. 
nuget.exe pack would already do weird things with an SDK project today. Customers that were observing the output would've moved away anyways as the warnings would be something that'd be difficult to fix. 

It's similar motivation as the whole, why not do a warning thing I mentioned in the alternatives:

>   Instead of an error, we could add a warning instead. Given that NuGet.exe pack by default will usually raise a bunch of warnings, it is likely that if warnings were good enough, the customers would have reviewed said warnings and figured out that all those warnings cannot be addressed correctly.